docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,139 @@
+
+<!--
+> TL;DR: Give a summary of what this document is proposing and what components it is touching.
+>
+> *For example: This design doc is proposing a consistent design template for “example.com” organization.*
+-->
+ 
+This design doc proposes to add tenancy awareness in the query path. 
+
+## Why
+
+<!--
+Provide a motivation behind the change proposed by this design document, give context.
+
+*For example: It’s important to clearly explain the reasons behind certain design decisions in order to have a consensus between team members, as well as external stakeholders. Such a design document can also be used as a reference and knowledge-sharing purposes. That’s why we are proposing a consistent style of the design document that will be used for future designs.*
+-->
+
+In a multi-tenant environment, it's important to be able to identify which tenants are experiencing issues and configure (e.g. with different limits) each one of them individually and according to their usage of the platform so that the quality of service can be guaranteed to all the tenants. 
+
+
+### Pitfalls of the current solution
+
+<!--
+What specific problems are we hitting with the current solution? Why it’s not enough?
+
+*For example, We were missing a consistent design doc template, so each team/person was creating their own. Because of inconsistencies, those documents were harder to understand, and it was easy to miss important sections. This was causing certain engineering time to be wasted.*
+-->
+
+The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools or proxies. For example, it's impossible to determine which tenants are experiencing high latency or high error rates.
+
+## Goals
+
+<!--
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow easy collaboration and decision making on design ideas.
+* Have a consistent design style that is readable and understandable.
+* Have a design style that is concise and covers all the essential information.
+-->
+
+* Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path. These features include, but aren't limited to, the following:
+  * Per-tenant observability.
+  * Per-tenant settings. For example, having different limits per tenant, which is a common request. 
+  * Enforce presence of one or more tenant labels in queries.
+
+### Audience
+
+<!--
+If this is not clear already, provide the target audience for this change.
+-->
+
+Any team running Thanos in a multi-tenant environment.
+
+## Non-Goals
+
+<!--
+* Move old designs to the new format.
+* Not doing X,Y,Z.
+-->
+
+* Add multi-tenancy to Thanos Ruler. It's not needed. One will still be able to point their Ruler to the multi-tenant Query Frontend, effectively making it multi-tenant. 
```

Comment:
It is needed, as rule configuration needs to be treated separately. You can select a series of a particular tenant, but you must also add back that label to the newly generated series, which makes it a bit more complex. Also, in stateless mode, Ruler needs per-tenant config to write to tenant TSDBs. 

Maybe let's skip mentioning need here, and just mention that it is out of scope for this proposal? 🙂 
```suggestion
* Add multi-tenancy to Thanos Ruler. 
```

Reply:
Rules can be already configured to add the tenant label, no @saswatamcode? For example, in our deployments we are putting the results of recording rules under specific tenants, otherwise it would be a complete mess. The write path is already prepared for this.

The proposal applies only the the query path and in reality there's a problem there. If the Ruler doesn't indicate a tenant through the HTTP header of a query, the default tenant will be picked, and then the tenant label enforcement and verification will fail because of the mismatch. **So in fact we need to update the ruler**.

Reply:
> So in fact we need to update the ruler.

Can we clarify off the back of this conversation what needs to be updated then?

Reply:
Alright, I did some investigation. We import the rule evaluation code directly from Prometheus, which owns the format of the rule configuration files and all the logic behind it. This makes fixing the Ruler a harder problem. We have two paths forward:

1. On a Querier with multi-tenancy enabled (`--querier.tenant-header` present), change the behavior when the header isn't sent. So instead of using the default tenant it would skip tenancy check. It seems like a bad idea to have a different behavior from Receive in the same situation (Receive considers requests without the tenant header to be meant for the default tenant). **This is not what I think we should do**.

2. Leave the Ruler out of the proposal. Meaning it will need to have its own pool of Queriers with multi-tenancy disabled to keep working normally. A new proposal can be created in the future to outline how we handle multi-tenancy in Ruler.

So, for now, I prefer to go with 2. 

WDYT, @saswatamcode?

Reply:
FYI I'll update the diagram to reflect my preference for option 2.

Reply:
Let's go with option 2 for now! :)

Reply:
> Can we clarify off the back of this conversation what needs to be updated then?

@PhilipGough Ruler would have to know to which tenant a given rule belongs to, so that when sending queries it includes that tenant's name in the tenant header. This means we would need to tap into the rule configuration format and evaluation engine, as both are  unaware of tenancy, as they belongs to Prometheus.

Reply:
btw, continuing on ⬆️ , Ruler would also need to know the tenant for the local (TSDB, requires no change) or remote (remote write, requires change to send tenant header) storage.